### PR TITLE
[IMP] mail: Mention user in chatter using @ functionality

### PR DESCRIPTION
**Pull request by jat-odoo**
- Task: https://www.odoo.com/web#id=9441&view_type=form&model=project.task&menu_id=3942&action=327
- Pad: https://pad.odoo.com/p/r.QEsd9DcnMk6WIY5v

### CODE REVIEW

File: addons/mail/static/src/js/mail.js [hunk 1]
```diff
@@ -333,6 +333,133 @@ openerp.mail = function (session) {
...
+              this.each(function(index, elem) {
+                if (elem.setSelectionRange)
+                  elem.setSelectionRange(pos, pos);
+                else if (elem.createTextRange)
+                  elem.createTextRange().collapse(true).moveEnd('character', pos).moveStart('character', pos).select();
+              });
+              return this;
+            };
+        },
+    }),
```
Comment (vme-odoo) [inline addons/mail/static/src/js/mail.js]: Need to check is this really needed or other options available.(myself)

File: addons/mail/static/src/js/mail.js [hunk 2]
```diff
@@ -333,6 +333,133 @@ openerp.mail = function (session) {
...
+                (last_ind != 0 && left.charAt(last_ind - 1) != " ")) return; // @ mention work if there is before space
+            this.search_str = left_str.slice(1);
+            if(this.search_str) {
+                this.timer = setTimeout(function() {
+                    self.start_index = index - self.search_str.length - 1;
+                    new openerp.web.Model("mail.compose.message").call("get_mention_users", [self.search_str]).then(function(res) {
+                        if(res.length)
+                            self.$el.html($(session.web.qweb.render('mail.mention', {'widget': _.extend(self, {"result": res})})).find("ul")).find("li:first").addClass("active");
+                    });
+                }, 1000);
```
Comment (vme-odoo) [inline addons/mail/static/src/js/mail.js]: searching code move in to new method.

File: addons/mail/static/src/js/mail.js [hunk 3]
```diff
@@ -649,6 +777,15 @@ openerp.mail = function (session) {
 
         on_message_post: function (event) {
             var self = this;
+            var textarea = self.$('textarea');
+            var body = textarea.val().replace(new RegExp("\n", "g"), "<br/>");
+            _.each(_.chain(this.mention_ids).keys().sortBy(function(k){ return -k.length; }).value(), function(key) {
+                var dict = self.mention_ids[key], word = _.str.sprintf("@%s ", key);
+                var reg_exp = new RegExp(word.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&"), 'g');
```
Comment (jat-odoo) [inline addons/mail/static/src/js/mail.js]: @vme-odoo done!

File: addons/mail/static/src/js/mail.js (hunk 2 above)
Comment (jat-odoo) [inline addons/mail/static/src/js/mail.js]: @vme-odoo 
From specification cache concept remove by the RIM

File: addons/mail/static/src/xml/mail.xml [hunk 4]
```diff
@@ -63,6 +63,22 @@
...
     <!--
+        mail.mention template
+        Dropdown list for @ mention functionality
+        -->
+    <t t-name="mail.mention">
+        <div class="mention">
+            <ul>
+                <t t-if="widget.result">
+                    <t t-foreach="widget.result" t-as='res'>
+                        <li><img height='25' width='25' t-attf-src='/web/binary/image?model=res.partner&amp;field=image_small&amp;id=#{res.partner_id[0]}' /> <span t-attf-id='#{res.partner_id[0]}'><t t-raw='widget.highlight(res.partner_id[1], widget.search_str)' /></span> (<i><t t-raw='widget.highlight(res.login, widget.search_str)' /></i>)</li>
```
Comment (jat-odoo) [inline addons/mail/static/src/xml/mail.xml]: @vme-odoo It's not css. height/width is attribute of  tag in html

File: addons/mail/static/src/js/mail.js [hunk 5]
```diff
@@ -544,6 +671,7 @@ openerp.mail = function (session) {
             this.$el.remove();
             this.$el = $render;
 
+            new mail.Mention(this).insertAfter(this.$el.find(".field_text")); // Initialize mail.Mention widget
```
Comment (jat-odoo) [inline addons/mail/static/src/js/mail.js]: @vme-odoo 
From reinit method i found textarea.

File: addons/mail/static/src/xml/mail.xml (hunk 4 above)
Comment (vme-odoo) [inline addons/mail/static/src/xml/mail.xml]: @jat-odoo are you sure its not inline? :)

Comment (jat-odoo) [inline addons/mail/static/src/xml/mail.xml]: @vme-odoo yes

File: addons/mail/static/src/js/mail.js (hunk 5 above)
Comment (vme-odoo) [inline addons/mail/static/src/js/mail.js]: this.mail_mention = new mail.Mention(this) like this you can store

File: addons/mail/static/src/js/mail.js [hunk 6]
```diff
@@ -333,6 +333,133 @@ openerp.mail = function (session) {
...
+                    this.set_text($active);
+                    return false;
+            }
+            this.$el.empty();
+            return true;
+        },
+        keyup: function(e) {
+            var self = this;
+            clearTimeout(this.timer);
+            if(_.contains([40, 38, 13, 37, 39, 35, 33, 34, 27], e.which)) { return; }
```
Comment (jat-odoo) [inline addons/mail/static/src/js/mail.js]: @vme-odoo done!

File: addons/mail/wizard/mail_compose_message.py [hunk 7]
```diff
@@ -384,3 +384,7 @@ def render_template(self, cr, uid, template, model, res_id, context=None):
 
     def render_message(self, cr, uid, wizard, res_id, context=None):
         return self.render_message_batch(cr, uid, wizard, [res_id], context)[res_id]
+
+    # Get mention user list (@ Functionality)
+    def get_mention_users(self, cr, uid, word, context=None):
+        return self.pool.get('res.users').search_read(cr, uid, ['|',('login','ilike', word), ('name','ilike', word)], ['partner_id', "login"], limit=8, context=context) if self.pool.get('res.users').has_group(cr, uid, 'base.group_user') else []
```
Comment (jat-odoo) [inline addons/mail/wizard/mail_compose_message.py]: @vme-odoo done!

File: addons/mail/static/src/js/mail.js (hunk 5 above)
Comment (jat-odoo) [inline addons/mail/static/src/js/mail.js]: @vme-odoo done!

File: addons/mail/static/src/js/mail.js [hunk 8]
```diff
@@ -333,6 +333,133 @@ openerp.mail = function (session) {
...
+     */
+
+    mail.Mention = session.web.Widget.extend({
+        template: 'mail.mention',
+        init: function(parent) {
+            this._super.apply(this, arguments);
+            this.parent = parent;
+            this.textarea = parent.$el.find(".field_text");
+            this.start_index = 0;
+            parent.mention_ids = {};
```
Comment (jat-odoo) [inline addons/mail/static/src/js/mail.js]: @vme-odoo done!

File: addons/mail/static/src/js/mail.js [hunk 9]
```diff
@@ -538,12 +676,20 @@ openerp.mail = function (session) {
...
         reinit: function() {
+            var self = this;
             var $render = $( session.web.qweb.render('mail.compose_message', {'widget': this}) );
 
             $render.insertAfter(this.$el.last());
             this.$el.remove();
             this.$el = $render;
 
+            // Initialize mail.Mention widget if current user is from employee group
+            new session.web.Model('res.users').call('has_group', ['base.group_user']).done(function(is_employee) {
```
Comment (vme-odoo) [inline addons/mail/static/src/js/mail.js]: Quick fixes @jat-odoo, but checking group rights in reinit method not gud, it shall be called only once.

File: addons/mail/wizard/mail_compose_message.py [hunk 10]
```diff
@@ -524,3 +524,7 @@ def render_message(self, cr, uid, wizard, res_id, context=None):
 
     def generate_email_for_composer(self, cr, uid, template_id, res_id, context=None):
         return self.generate_email_for_composer_batch(cr, uid, template_id, [res_id], context)[res_id]
+
+    # Get mention user list (@ Functionality)
+    def get_mention_users(self, cr, uid, word, context=None):
+        return self.pool.get('res.users').search_read(cr, uid, ['|', ('login', 'ilike', word), ('name', 'ilike', word)], ['partner_id', 'login'], limit=8, context=context)
```
Comment (vme-odoo) [inline addons/mail/wizard/mail_compose_message.py]: whats meaning of *word* here? and limit shall be pass from js, where we can configure it.

**Comment by vme-odoo** [comment]
let me check and commit some issue my self, thanks @jat-odoo.

File: addons/mail/wizard/mail_compose_message.py (hunk 10 above)
Comment (jat-odoo) [inline addons/mail/wizard/mail_compose_message.py]: @vme-odoo word mean search term (text after @ type)

Comment (vme-odoo) [inline addons/mail/wizard/mail_compose_message.py]: typo was not good, please make it understandable next time.

Comment (jat-odoo) [inline addons/mail/wizard/mail_compose_message.py]: @vme-odoo done!

**Comment by jholze** [comment]
Greate job :+1:

File: addons/mail/mail_followers.py [hunk 11]
```diff
@@ -191,7 +191,7 @@ def _notify_email(self, cr, uid, ids, message_id, force_send=False, user_signatu
             return True
 
         # compute email body (signature, company data)
-        body_html = message.body
+        body_html = html_sanitize(context.get('mail_body')) if context.get('mail_body') else message.body
```
Comment (rim-odoo) [inline addons/mail/mail_followers.py]:
```python
context.get('mail_body', message.body)
```

File: addons/mail/wizard/mail_compose_message.py [hunk 12]
```diff
@@ -528,3 +528,7 @@ def render_message(self, cr, uid, wizard, res_id, context=None):
 
     def generate_email_for_composer(self, cr, uid, template_id, res_id, context=None):
         return self.generate_email_for_composer_batch(cr, uid, template_id, [res_id], context)[res_id]
+
+    # Get mention user list (@ Functionality)
+    def get_mention_users(self, cr, uid, search_str, limit, context=None):
```
Comment (rim-odoo) [inline addons/mail/wizard/mail_compose_message.py]: maybe add a default limit?

File: addons/mail/wizard/mail_compose_message.py [hunk 13]
```diff
@@ -528,3 +528,7 @@ def render_message(self, cr, uid, wizard, res_id, context=None):
...
+        return self.pool.get('res.users').search_read(cr, uid, ['|', ('login', 'ilike', search_str), ('name', 'ilike', search_str)], ['partner_id', 'login'], limit=limit, context=context)
```
Comment (rim-odoo) [inline addons/mail/wizard/mail_compose_message.py]: on res_users object, name_search function already searches by login then by name

File: addons/mail/wizard/mail_compose_message.py (hunk 12 above)
Comment (jat-odoo) [inline addons/mail/wizard/mail_compose_message.py]: @rim-odoo done!

File: addons/mail/wizard/mail_compose_message.py (hunk 13 above)
Comment (jat-odoo) [inline addons/mail/wizard/mail_compose_message.py]: @rim-odoo 
But we need partner_id, name and login fields.
while name_search just return user id and name.

File: addons/mail/mail_followers.py (hunk 11 above)
Comment (jat-odoo) [inline addons/mail/mail_followers.py]: @rim-odoo 
If 'mail_body' in context available than we have to use html_sanitize.
So, Use <code>html_sanitize(context.get('mail_body')) or message.body</code>

**Comment by rim-odoo** [comment]
@jat-odoo Can you rebase your code with latest master, there are some conflicts? Can you also check that recent changes in our JavaScript framework did not break anything? Thanks a lot

**Comment by hmo-odoo** [comment]
new PR: https://github.com/odoo/odoo/pull/7225

**Pull request closed without merging**